framework/ui/view/launchprovider.cpp

```diff
@@ -137,7 +163,66 @@ RetVal<Val> QmlLaunchProvider::toRetVal(const QVariant& jsrv) const
     return rv;
 }
 
-void QmlLaunchProvider::onOpen(QString pageType)
+RetVal<LaunchProvider::OpenData> LaunchProvider::openWidget(const UriQuery &q)
+{
+    RetVal<OpenData> result;
+
+    QString uri = QString::fromStdString(q.uri().toString());
+
+    ContainerMeta container = uriRegister()->container(uri);
+    int widgetMetaTypeId = container.widgetMetaTypeId;
+
+    QString objectId = QString::number(widgetMetaTypeId); // unque?
+
+    void *widgetClassPtr = QMetaType::create(widgetMetaTypeId);
+    QDialog* dialog = static_cast<QDialog*>(widgetClassPtr);
+
+    if (!dialog) {
+        result.ret = make_ret(Ret::Code::UnknownError);
+        return result;
+    }
+
+    fillData(dialog, q);
+
+    QVariantMap status;
+    connect(dialog, &QDialog::finished, [=, &status, &objectId](int finishStatus){
+        status["errcode"] = static_cast<int>(Ret::Code::Ok);
+        status["value"] = finishStatus;
+
+        onPopupClose(objectId, status);
+    });
+
+    onOpen(PAGE_TYPE_WIDGET);
+
+    dialog->exec();
+
+    QMetaType::destroy(widgetMetaTypeId, widgetClassPtr);
+
+    result.ret = toRet(status);
+    result.val.sync = false; // TODO
+    result.val.objectID = QString::number(widgetMetaTypeId);
+
+    return result;
+}
+
+RetVal<LaunchProvider::OpenData> LaunchProvider::openQml(const UriQuery &q)
+{
+    QmlLaunchData* data = new QmlLaunchData();
+    fillData(data, q);
+
+    emit fireOpen(data);
+
+    RetVal<OpenData> result;
+    result.ret = toRet(data->value("ret"));
+    result.val.sync = data->value("sync").toBool();
+    result.val.objectID = data->value("objectID").toString();
+
+    delete data;
+
+    return result;
+}
+
+void LaunchProvider::onOpen(QString pageType)
```

Comment:
Now this method needs some work, you need:
* change argument from QString to ContainerType
* remove constants PAGE_TYPE_ *